### PR TITLE
feature: NotarizedTransactionV2 can be validated

### PR DESCRIPTION
**Pull request by dhedey**
## Summary

Main changes:
* Implements `TransactionValidator` for `NotarizedTransactionV2`, including all the behaviour from the SubIntent model specification.
* By default, `ManifestBuilder.build()` validates all built manifests with the new/more advanced validator. This can be opted out of with `build_no_validate()`.
* Fixes some bugs with the V2 transaction builders, and in the validator.
* Adds a scenario which runs subintents which we can use for Cuttlefish.

Up next:
* Resolve the config for the transaction builder
* Get Josh's PR merged
* Loads of tests to come next, as per the test list

## Testing
* Existing tests pass
* Lots of tests to come

### CODE REVIEW

File: radix-transactions/src/errors.rs [hunk 1]
```diff
@@ -99,3 +140,14 @@ pub enum InvalidMessageError {
         curve_type: CurveType,
     },
 }
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SubintentValidationError {
```
Comment (dhedey) [inline radix-transactions/src/errors.rs:145]: I will write tests in follow up PRs to cover all of these error possibilities

File: radix-transactions/src/validation/transaction_validator.rs [hunk 2]
```diff
@@ -379,12 +467,646 @@ impl TransactionValidator {
...
+        // STEP 3
+        // ------
+        // We traverse the child relationships from the root, and mark a depth.
+        // We error if any exceed the maximum depth.
+        //
+        // As each child has at most one parent, we can guarantee the work is bounded
+        // by the total number of subintents.
+        let mut work_list = vec![];
+        for index in root_intent_details.children.iter() {
+            work_list.push((*index, 1));
```
Comment (iamyulong) [inline radix-transactions/src/validation/transaction_validator.rs:928]: depth typically start with `0` at the root.

File: radix-transactions/src/validation/transaction_validator.rs [hunk 3]
```diff
@@ -379,12 +467,646 @@ impl TransactionValidator {
...
+            subintents: all_subintent_details,
+        })
+    }
+}
+
+struct HeaderAggregation {
+    overall_start_epoch_inclusive: Epoch,
+    overall_end_epoch_exclusive: Epoch,
+    overall_start_timestamp_inclusive: Option<Instant>,
+    overall_end_timestamp_exclusive: Option<Instant>,
```
Comment (iamyulong) [inline radix-transactions/src/validation/transaction_validator.rs:977]: Do we intend to validate the overall timestamp range?

File: radix-transactions/src/validation/transaction_validator.rs [hunk 4]
```diff
@@ -379,12 +467,646 @@ impl TransactionValidator {
...
+                &subintent.intent_core,
+                &mut header_aggregation,
+                true,
+            )?;
+            yield_summaries.insert(subintent.subintent_hash().into(), yield_summary);
+        }
+
+        for (child_hash, child_details) in subintents_details {
+            let child_intent_hash = IntentHash::Subintent(*child_hash);
+            // The instruction validation checks that the
```
Comment (iamyulong) [inline radix-transactions/src/validation/transaction_validator.rs:585]: Comment is half done?

File: radix-transactions/src/validation/transaction_validator.rs [hunk 5]
```diff
@@ -379,12 +467,646 @@ impl TransactionValidator {
...
+        if header.tip_basis_points < self.config.min_tip_basis_points
+            || header.tip_basis_points > self.config.max_tip_basis_points
+        {
+            return Err(HeaderValidationError::InvalidTip);
+        }
+
+        Ok(())
+    }
+
+    fn validate_v2_intent_cores(
```
Comment (iamyulong) [inline radix-transactions/src/validation/transaction_validator.rs:554]: May worth noting that `validate_v2_intent_cores` assumes the transaction has passed `validate_intent_relationships_v2`.

**Review by iamyulong — COMMENTED** [review]
Nice work!

Although, I feel a bit overwhelmed with the names and would suggest the following:
```
transaction_intent  (root_subintent)
root_intent         (root_subintent)
subintents          (non_root_subintents)
child_subintents    (non_root_subintents)
all_subintents      (all_subintents)
is_subintent        (is_root_subinent)
parent              (parent)              // describe local relationship
child_subintent     (child)               // describe local relationship

intent = vec<subintent> 
       = root_subintent + non_root_subintents
```

File: radix-transactions/src/validation/transaction_validator.rs (hunk 2 above)
Comment (dhedey) [reply radix-transactions/src/validation/transaction_validator.rs:928]: I agree - these are the depths for the root's children (i.e. the first layer of subintents)

File: radix-transactions/src/validation/transaction_validator.rs (hunk 3 above)
Comment (dhedey) [reply radix-transactions/src/validation/transaction_validator.rs:977]: I believe this is validated in `HeaderAggregation::update`? Is there some other validation you've noticed that I've missed?

File: radix-transactions/src/validation/transaction_validator.rs (hunk 5 above)
Comment (dhedey) [reply radix-transactions/src/validation/transaction_validator.rs:554]: True.

It takes a subintent relationship details argument which only comes from validating the relationships - but I'll add a comment clarifying :).

**Review by iamyulong — APPROVED** [review]
Rename will be done in a separate PR.

**Pull request merged**